allennlp/models/semantic_parsing/wikitables/wikitables_semantic_parser.py

```diff
@@ -518,133 +507,154 @@ def get_metrics(self, reset: bool = False) -> Dict[str, float]:
+            # `key` here is a non-terminal from the grammar, and `action_strings` are all the valid
+            # productions of that non-terminal.  We'll first split those productions by global vs.
+            # linked action.
+            action_indices = [action_map[action_string] for action_string in action_strings]
+            production_rule_arrays = [(possible_actions[index], index) for index in action_indices]
+            global_actions = []
+            linked_actions = []
+            for production_rule_array, action_index in production_rule_arrays:
+                if production_rule_array[1]:
+                    global_actions.append((production_rule_array[2], action_index))
+                else:
+                    linked_actions.append((production_rule_array[0], action_index))
+
+            # Then we get the ebmedded representations of the global actions.
+            global_action_tensors, global_action_ids = zip(*global_actions)
+            global_action_tensor = torch.cat(global_action_tensors, dim=0)
+            global_input_embeddings = self._action_embedder(global_action_tensor)
+            if self._add_action_bias:
+                global_action_biases = self._action_biases(global_action_tensor)
+                global_input_embeddings = torch.cat([global_input_embeddings, global_action_biases], dim=-1)
+            global_output_embeddings = self._output_action_embedder(global_action_tensor)
+            translated_valid_actions[key]['global'] = (global_input_embeddings,
+                                                       global_output_embeddings,
+                                                       list(global_action_ids))
+
+            # Then the representations of the linked actions.
+            if linked_actions:
+                linked_rules, linked_action_ids = zip(*linked_actions)
+                entities = [rule.split(' -> ')[1] for rule in linked_rules]
+                entity_ids = [entity_map[entity] for entity in entities]
+                # (num_linked_actions, num_question_tokens)
+                entity_linking_scores = linking_scores[entity_ids]
+                # (num_linked_actions,)
+                entity_type_tensor = entity_types[entity_ids]
+                # (num_linked_actions, entity_type_embedding_dim)
+                entity_type_embeddings = self._entity_type_decoder_embedding(entity_type_tensor)
+                translated_valid_actions[key]['linked'] = (entity_linking_scores,
+                                                           entity_type_embeddings,
+                                                           list(linked_action_ids))
+
+        # Lastly, we need to also create embedded representations of context-specific actions.  In
+        # this case, those are only variable productions, like "r -> x".
+        context_actions = {}
+        for action_id, action in enumerate(possible_actions):
+            if action[0].endswith(" -> x"):
```

Comment:
This will break if we had nested lambdas, right? Wouldn't happen with WikiTables, but may be add a TODO?

Reply:
The `WikiTablesWorld` specifies that there is at most one nested lambda: https://github.com/allenai/allennlp/blob/75abebb437042e4709ecf4e9af770df6a4912896/allennlp/semparse/worlds/wikitables_world.py#L43

I'll add a note here, not a TODO.